New README (#21)
aaronwolen/osfr: README.Rmd

---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)

osfr2:::osf_dev_off()
```

# osfr2

[![Build Status](https://travis-ci.org/aaronwolen/osfr.svg?branch=master)](https://travis-ci.org/aaronwolen/osfr)
[![Coverage status](https://codecov.io/gh/aaronwolen/osfr/branch/master/graph/badge.svg)](https://codecov.io/github/aaronwolen/osfr?branch=master)

## Overview

osfr provides a suite of functions for interacting with [OSF][osf] that are primarily focused on project management workflows.

> OSF (Open Science Framework, <https://osf.io>) is a free and [open source][osf-gh] project management repository designed to support researchers of all technical backgrounds. The service includes unlimited cloud storage and file version history, providing a centralized location for your research materials that can be kept private, shared with select collaborators, or made publicly available with citable DOIs.

Reviewer: ```suggestion
OSF (<https://osf.io>) is a free and [open source][osf-gh] project management repository designed to support researchers of all technical backgrounds. The service includes unlimited cloud storage and file version history, providing a centralized location for all your research materials that can be kept private, shared with select collaborators, or made publicly available with citable DOIs.
```

Author: Is "open science framework" the new Voldemort?  🙂

## Installation

This package is currently under development and is apt to undergo significant changes until a stable version has been submitted to CRAN. You've been warned.

Also note this version of osfr has been temporarily renamed to *osfr2*.

``` r
# install.packages("devtools")
devtools::install_github("aaronwolen/osfr")
```

## Getting started

There are 3 main types of OSF entities that osfr can work with:

1. **nodes:** both [projects][help-proj] and [components][help-comp] (i.e., sub-projects) are referred to as nodes
2. **files:** this includes both files *and* folders stored on OSF
3. **users:** individuals with OSF accounts

osfr represents these entities within `osf_tbl`s---specialized tibbles that provide  useful information about the entities like their `name` and unique `id` for users, and API data in the `meta` column that's necessary for osfr's internal functions. Otherwise, they're just `data.frames` and can be manipulated using standard functions from base R or dplyr.

## Usage examples

### Accessing Open Research Materials

> Many researchers use OSF to archive and share their work. If you come across a paper that cites an OSF repository, osfr can be used to explore the project and download the associated files---you only need the project's URL or GUID to get started. In this case, we're retrieving the Cancer Reproducibility Project (<https://osf.io/e81xl/>):

Reviewer: ```suggestion
Many researchers use OSF to archive and share their work. If you come across a paper that cites an OSF repository, osfr can be used to explore the project and download the associated files---you only need the project's URL or GUID (global unique identifier) to get started. In this case, we're retrieving the Cancer Reproducibility Project (<https://osf.io/e81xl/>):
```

```{r message=FALSE}
library(osfr2)

cr_project <- osf_retrieve_node("e81xl")
cr_project
```

This returns an `osf_tbl_node` object with 1 row for the OSF project we retrieved. Let's list the files that have been uploaded to this project:

```{r}
cr_files <- osf_ls_files(cr_project)
cr_files
```

> This returns an `osf_tbl_file` with 1 row for each of the uploaded files. We can examine any of these files directly on OSF with `osf_open()`, which opens the corresponding file's view in your default browser.

Reviewer: ```suggestion
This returns an `osf_tbl_file` with 1 row for each of the files stored in osf-storage. OSF allows for other resources, like GitHub repos and dropbox folders, to be connected to projects, but currently the osfr package only indexes files stored locally in osf-storage. We can examine any of these files directly on OSF with `osf_open()`, which opens the corresponding file's view in your default browser.
```

```{r}
my_project <- osf_create_project(title = "Motor Trend Car Road Tests")
my_project
```

Using a combination of `osf_create_component()`, `osf_mkdir()`, and `osf_upload()`, you can easily implement your preferred organizational structure for a project and populate it with files. Note that osfr's functions are [pipe-friendly][magrittr], meaning they can be arranged together in pipelines:

```{r}
library(magrittr)
# create a local copy of mtcars
write.csv(mtcars, "mtcars.csv")

# upload to the rawdata folder in the "Car Data" project component
my_project %>%
  osf_create_component("Car Data") %>%
  osf_mkdir("rawdata") %>%
  osf_upload("mtcars.csv") %>%
  osf_open()
```

![Screenshot of the uploaded file on OSF](man/figures/screen-shot.png)

```{r cleanup, include=FALSE}
unlink("mtcars.csv")
osf_rm(my_project, recursive = TRUE, verbose = FALSE)
```

## Acknowledgments

OSF is developed by the [Center for Open Science][cos] in Charlottesville, VA.

> The original version of osfr was developed by @chartgerink. The current version was developed by @aaronwolen and is *heavily* inspired by @jennybc and @lucymcgowan's excellent [googledrive][] package. Seriously, we borrowed a lot of great ideas from them. Other important resources include [http testing](https://ropensci.github.io/http-testing-book/) by Scott Chamberlain and [R Packages](http://r-pkgs.had.co.nz) by Hadley Wickham. Development was also greatly facilitated by OSF's excellent [API documentation][osf-api].

Reviewer: ```suggestion
The original version of osfr was developed by @chartgerink and further developed by @bgrich. The current version was developed by @aaronwolen and is *heavily* inspired by @jennybc and @lucymcgowan's excellent [googledrive][] package. Seriously, we borrowed a lot of great ideas from them. Other important resources include [http testing](https://ropensci.github.io/http-testing-book/) by Scott Chamberlain and [R Packages](http://r-pkgs.had.co.nz) by Hadley Wickham. Development was also greatly facilitated by OSF's excellent [API documentation][osf-api].
```
